### PR TITLE
feat: added Types module

### PR DESCRIPTION
**Pull request by jessekelly881**
Adds a collection of types used across effect repos

### CODE REVIEW

**Comment by mikearnaldi** [comment]
I see where you're coming from as objects with a tag are very frequent but I am a bit reluctant to add a module for the sake of two type helpers that are basically an alias of Extract / Exclude. Namely whenever we had to deal with tags it is to do different things, almost never a simple exclusion or similar so I guess to bring this forward we'd need a strong motivation with concrete examples of app code.

Alternatively: we do have a number of type helpers that really have no place and we've been discussing the creation of a Types module to put there commonly used types such as IsEqual / UnionToIntersection & co (cc @tim-smart / @gcanti)

**Comment by jessekelly881** [comment]
I'm inclined to agree with you. My first thought was actually to create a Types module and only went with a Tagged module because the types that I wanted to add were only specific to tagged objects.

**Comment by gcanti** [comment]
in my TODO list there's an old "move UnionToIntersection to some centralized location", +1 for a Types module

**Comment by jessekelly881** [comment]
Converted to Types module.

**Comment by gcanti** [comment]
@jessekelly881 There's also `type Simplify<A>` in `src/Data` (used for example in /schema).

However I would change it to:

```ts
export type Simplify<A> = {
  [K in keyof A]: A[K]
} extends infer B ? B : never
```

which IME produces shorter types in the emitted declaration files

**Comment by jessekelly881** [comment]
I didn't see that specific type-def (I'm probably missing just overlooking it). I added Simplify and moved Data/IsEqualTo -> Types.Equal

**Comment by gcanti** [comment]
> I didn't see that specific type-def

you mean Simplify? is here https://github.com/Effect-TS/data/blob/62d6dc25f02dc489ce98399ae661fabdd00833e0/src/Data.ts#L191
(I guess you are a few commits behind `main`)

**Comment by jessekelly881** [comment]
I was indeed a few commits behind main. lol. Updated. :)

**Comment by jessekelly881** [comment]
I was indeed a few commits behind. Lol.

File: dtslint/Data.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import * as Data from"@effect/data/Data";
+
+// $ExpectType Data.Data<>
```
Comment (gcanti) [inline dtslint/Data.ts:3]: this is raising an error in the `dtslint` task (not sure what's the purpose of this file though)

**Review by gcanti — CHANGES_REQUESTED** [review]

File: dtslint/Data.ts (hunk 1 above)
Comment (jessekelly881) [reply dtslint/Data.ts:3]: There isn't a point. Jaja This was something I was testing that slipped through. Let me clean this PR up(it's been neglected) and I will ping you in a bit when it's ready

Comment (jessekelly881) [reply dtslint/Data.ts]: Fixed. :)

**Comment by jessekelly881** [comment]
Are there any other types that you guys can think of that we should add to this?

**Review by gcanti — APPROVED** [review]

**Comment by tim-smart** [comment]
From /io:

image

**Comment by mikearnaldi** [comment]
LGTM @tim-smart if you want to push MergeRecord we can merge this

**Comment by jessekelly881** [comment]
I added it. I renamed it Type.Extend after S.extend and (at least to) me seems more clear. Any quarrels w/ that name instead?

**Comment by tim-smart** [comment]
> I added it. I renamed it to Type.Extend after S.extend and (at least to me) that seems more clear. Any quarrels w/ that name instead?

I think if it is called Extend, the second type parameter should have priority when it comes to key conflicts.

**Comment by jessekelly881** [comment]
Make sense. Do you think it would be better to change the functionality or the name then?

**Comment by tim-smart** [comment]
I think MergeRight and MergeLeft is good.

**Comment by tim-smart** [comment]
@jessekelly881 if you can run `pnpm run docs` and then we can merge this :)

**Pull request merged**